generic/pgtclId.c

```diff
@@ -83,12 +84,19 @@ PgInputProc(DRIVER_INPUT_PROTO)
 
 	/* Move data from libpq's buffer to Tcl's. */
 
-	avail = PQgetlineAsync(conn, buf, bufSize);
+	switch(avail = PQgetCopyData(conn, &bufPtr, bufSize)) {
+	  case -2: {
+		*errorCodePtr = EIO;
+		return -1;
+	  }
+	  case -1: {
+		return PgEndCopy(connid, errorCodePtr, 0);
+	  }
+	}
 
-	if (avail < 0)
-	{
-		/* Endmarker detected, change state and return 0 */
-		return PgEndCopy(connid, errorCodePtr);
+	if(bufPtr) {
```

Comment:
Probably still need to ensure avail is a sane size_t value here before passing to memcpy.

Reply:
avail can be -2, -1, or 0 or greater, and -2 and -1 have been resolved in 87..95

Reply:
Added check for unexpected negative values of avail.